src/Carnac.Logic/MessageController.cs

```diff
@@ -60,49 +67,29 @@ public void Start()
                         .Switch()
                         .Select(_ => message)
                         .Take(1);
-                })
-                .Select(m => Tuple.Create(m, ActionType.FadeOut));
+                });
+            var fadeOutMessageSubscription = fadeOutMessageStream
+                .ObserveOn(concurrencyService.MainThreadScheduler)
+                .SubscribeOn(concurrencyService.MainThreadScheduler)
+                .Subscribe(m => m.IsDeleting = true);
 
             // Finally we just put a one second delay on the messages from the fade out stream and flag to remove.
             var removeMessageStream = fadeOutMessageStream
                 .Delay(OneSecond, concurrencyService.Default)
-                .Select(m => Tuple.Create(m.Item1, ActionType.Remove));
-            
-            var actionStream = Observable.Merge(
-                removeMessageStream, 
-                fadeOutMessageStream, 
-                addMessageStream);
-
-            actionSubscription = actionStream
                 .ObserveOn(concurrencyService.MainThreadScheduler)
-                .SubscribeOn(concurrencyService.MainThreadScheduler) // Because we mutate message state we need to do everything on UI thread. 
-                                                             // If we introduced a 'Update' action to this feed we could remove mutation from the stream
-                .Subscribe(action =>
-            {
-                switch (action.Item2)
-                {
-                    case ActionType.Add:
-                        keys.Add(action.Item1);
-                        break;
-                    case ActionType.Remove:
-                        keys.Remove(action.Item1);
-                        break;
-                    case ActionType.FadeOut:
-                        action.Item1.IsDeleting = true;
-                        break;
-                    default:
-                        throw new ArgumentOutOfRangeException();
-                }
-            }, ex =>
-            {
-                
-            });
+                .SubscribeOn(concurrencyService.MainThreadScheduler)
+                .Subscribe(m => messages.Remove(m));
 
+            stopCarnacDisposable = new CompositeDisposable(
+                messageStream.Connect(), 
+                addMessageSubscription, 
+                fadeOutMessageSubscription,
+                removeMessageStream);
         }
 
         public void Dispose()
         {
-            actionSubscription.Dispose();
+            if (stopCarnacDisposable != null) stopCarnacDisposable.Dispose();
```

Comment:
why not make the stopCarnacDisposable a SerialDisposable?  No null checking and just dispose it